ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs

```diff
@@ -0,0 +1,184 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NamedFieldPuns   #-}
+{-# LANGUAGE RankNTypes       #-}
+
+module Test.Consensus.Mempool (tests) where
+
+import           Test.QuickCheck (Property, Testable, classify, counterexample,
+                     tabulate, withMaxSuccess, (===))
+import           Test.Tasty (TestTree, testGroup)
+import           Test.Tasty.QuickCheck (testProperty)
+
+import           Control.Monad.Class.MonadSTM (atomically)
+import           Control.Monad.IOSim (SimM, runSimOrThrow)
+
+import           Ouroboros.Consensus.Ledger.Abstract (ledgerConfigView)
+import           Ouroboros.Consensus.Mempool (Mempool (..),
+                     MempoolSnapshot (..), openMempool)
+import           Ouroboros.Consensus.Mempool.TxSeq (TicketNo)
+import           Ouroboros.Consensus.Util.ThreadRegistry (withThreadRegistry)
+
+import           Ouroboros.Network.Chain (Chain (..))
+
+import           Ouroboros.Storage.ChainDB.API (ChainDB)
+import qualified Ouroboros.Storage.ChainDB.API as ChainDB
+import qualified Ouroboros.Storage.ChainDB.Mock as Mock
+
+import           Test.Util.TestBlock (GenTx (..), GenTxId (..), TestBlock,
+                     computeGenTxId, singleNodeTestConfig, testInitExtLedger)
+import           Test.Util.TestTx (TestTx (..))
+
+tests :: TestTree
+tests = testGroup "Mempool"
+  [ testProperty "getTxs == getTxsAfter zeroIdx"        prop_Mempool_getTxs_getTxsAfter
+  , testProperty "addedTxs == getTxs"                   prop_Mempool_addTxs_getTxs
+  , testProperty "Invalid transactions are never added" prop_Mempool_InvalidTxsNeverAdded
+  ]
+
+{-------------------------------------------------------------------------------
+  Mempool Implementation Properties
+-------------------------------------------------------------------------------}
+
+-- | Test that @getTxs == getTxsAfter zeroIdx@.
+prop_Mempool_getTxs_getTxsAfter :: Chain TestBlock
+                                -> [TestTx]
+                                -> Property
+prop_Mempool_getTxs_getTxsAfter bc txs = withMaxSuccess 1000 $
+  tabulate "Transactions" (map constrName txs) $
+  runSimOrThrow test
+ where
+  test :: forall s. SimM s Property
```

Comment:
I would use the most general type here, i.e. `MonadSTM m => m Property` (and any other required constraints) here (and throughout this module). The type will be instantiated by `runSimOrThrow` anyway.